usage/app/controllers/UsageApi.scala

```diff
@@ -321,19 +322,40 @@ class UsageApi(
       "requestId" -> RequestLoggingFilter.getRequestId(req),
       "image-id" -> mediaId,
     )
-    usageTable.queryByImageId(mediaId).map(usages => {
-      usages.foreach(usageTable.deleteRecord)
-    }).recover {
-      case error: BadInputException =>
-        logger.warn(logMarker, "UsageApi returned an error.", error)
-        respondError(BadRequest, "image-usage-delete-failed", error.getMessage)
-      case error: Exception =>
-        logger.error(logMarker, "UsageApi returned an error.", error)
-        respondError(InternalServerError, "image-usage-delete-failed", error.getMessage)
-    }
 
-    val updateMessage = UpdateMessage(subject = DeleteUsages, id = Some(mediaId))
-    notifications.publish(updateMessage)
+    var photoSalesUsage = false
+    var filteredUsages = ListBuffer[MediaUsage]()
+
+    usageTable.queryByImageId(mediaId)
+      .map(usages => {
+        usages.foreach(usage => {
+          if (usage.syndicationUsageMetadata.exists(_.partnerName == "Capture")) {
+            photoSalesUsage = true
+          } else {
+            filteredUsages.+=(usage)
+            usageTable.deleteRecord(usage)
+          }
+        })
+        if (!photoSalesUsage) {
+          val updateMessage = UpdateMessage(subject = DeleteUsages, id = Some(mediaId))
+          notifications.publish(updateMessage)
+        } else {
+          filteredUsages.foreach(usage => {
+            usageTable.deleteRecord(usage)
+            val usageId = Some(usage.grouping + "_" + usage.usageId.toString)
+            val updateMessage = UpdateMessage(subject = DeleteSingleUsage, id = Some(mediaId), usageId = usageId)
+            notifications.publish(updateMessage)
+          })
+        }
```

Comment:
Does this section of code line 339 - line 349 need to sit within the map command - in the original method the thrall elements sat outside that block and I think that could be the case here too?

Reply:
At first I tried to replicate the original approach, with the Thrall elements sitting outside the query block.

This resulted in an issue where the thrall elements were processed before the query block (despite coming after the query block in terms of lines of code) - this is because the query block invovles Futures.

I think this was already happening in the original code, but it wasn't picked up because the deletion from elastic search and the deletion from dynamoDB could occur independently of each other in this case. However for our case, we rely on logic from the query block to determine what usages should be removed from elastic. Therefore, with the thrall block executing first, it was resulting in all usages being deleted, and the photo sales usage was not retained.

By moving the thrall block inside the query block, this forced correct order of operation and provides the desired result.

Reply:
Okay